### PR TITLE
Matching kernel's function types with actual types.

### PR DESCRIPTION
**Pull request by zahiraam**
Using getFullyQualifiedType to calculcate the kernel function's types makes the
types in the kernel different from the actual type and that creates a type
mismatch.
getFullyQualifiedType's use is restricted to make type lookup correct in
templates.
This change, replacing the fully qualified type with the canonical
one, fixes the issue, but will imply that the pretty-printer needs to be
updated to print fully qualified names.

Signed-off-by: Zahira Ammarguellat <zahira.ammarguellat@intel.com>

### CODE REVIEW

**Comment by erichkeane** [comment]
I'd very much like a better structured commit message (the title line doesn't do a good job).

Additionally, I would have assumed that this would break a number of integration header tests, did it not, or did you not test yet?

**Comment by zahiraam** [comment]
> I'd very much like a better structured commit message (the title line doesn't do a good job).
> 
> Additionally, I would have assumed that this would break a number of integration header tests, did it not, or did you not test yet

**Comment by zahiraam** [comment]
> I'd very much like a better structured commit message (the title line doesn't do a good job).
> 
> Additionally, I would have assumed that this would break a number of integration header tests, did it not, or did you not test yet?

It is in the process of being tested. I realized that doing a check-all on debug build had some issues, building the prod version and then testing.

**Comment by erichkeane** [comment]
As an explanation for those reviewing:

The getFullyQualifiedType is intended to be used in phase1-template parsing code in order to replace known types with one that will be properly looked up during the 2nd phase of template lookup.  Basically:

    namespace Top {
      struct b;
      namespace Inside {
        template<typename T>
        void foo() {
          b some_var;
        }
        struct b; // Note this is not the same as the one above, it is in a different namespace!
        void bar() { foo<int>(); }
      }
     }

At the instantiation of 'foo', lookup is done at that spot, however the standard requires that we pick up the out version of 'b' in Top, not the second one. SO, we replace the version of 'b' in the uninstantiated template with an ElaboratedType that contains the fully qualified name (to ensure we find the right one).  These types are not intended to leave template instantiation.

However, when we were early in SYCL development on this repo, someone found we also needed the Fully Qualified Name (FQN) in the integration headers, and opted to implement this using the TypeName::getFullyQualifiedType function to change the kernel type.  This results in the type in the kernel instantiation being the ElaboratedType (a different type) from the one named later (since we do it so late it doesn't undergo canonicalization).

The correct solution here is to just use the canonical type (as this patch does) rather than attempt to insert the FQN this late.  The integration header code likely needs to ensure it just prints the FQN (which I believe we do in a number of other cases).

File: clang/test/SemaCXX/kernel-function-type.cpp [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -fsyntax-only -verify %s
+// expected-no-diagnostics
+struct b{};
```
[the remark this answers is not in the document]
Comment (erichkeane) [reply clang/test/SemaCXX/kernel-function-type.cpp:3]: SemaCXX is probably the wrong place for this...  This is a minimization that we did, so I'll do my best to explain what happened.

Basically, the kernel_single_task call was emitted as an OpenCL kernel.  Our call to getFullyQualifiedType caused a now-2nd instantiation of zip_iterator<b,b> (first was on line 16 during phase1).

Then, the call to foo  in main causes 'foo' to be instantiated.  The best match for zip_iterator<j,d> is actually now the one instantiated by the kernel_single_task, which is now different from the one made in phase 1.

Thus is_same<zip_iterator<b,b>, zip_iterator<j,d>>::value is not true, despite both being zip_iterator<b,b>.

This is because it is ACTUALLY:

is_same<zip_iterator<b,b> (as instantiated during phase 1 of line 16), zip_iterator<b,b>(instantiated by getFullyQualifiedName)>.

[the remark this answers is not in the document]
Comment (zahiraam) [reply clang/test/SemaCXX/kernel-function-type.cpp:3]: This is the test case (long) that is failing. In this test the assert shouldn't go off, as it is supposed to test two same types. But in the current compiler (without this patch), it is failing in the assert.
#include <CL/sycl.hpp>

#include <dpstd/iterators.h>

namespace sycl = cl::sycl;

namespace dpct {
  namespace internal {
    // struct for checking if iterator is heterogeneous or not
    template <typename Iter, typename Void = void> // for non-heterogeneous iterators
            struct is_hetero_iterator
      : std::false_type
    { };

    template <typename Iter> // for heterogeneous iterators
    struct is_hetero_iterator<Iter, typename std::enable_if<Iter::is_hetero::value, void>::type>
      : std::true_type
    { };
  }
  template <typename T/*,*/
  /*sycl::access::mode Mode = sycl::access::mode::read_write*//*,
    typename Allocator = sycl::buffer_allocator*/>
  class device_iterator {
  protected:
    sycl::buffer<T, 1/*, Allocator*/> buffer;
    std::size_t idx;

  public:
    using value_type = T;
    using difference_type = std::make_signed<std::size_t>::type;
    using pointer = T*;
    using reference = T&;
    using iterator_category = std::random_access_iterator_tag;
    using is_hetero = std::true_type;                    // required
    static constexpr auto /*sycl::access::mode*/ mode =
      cl::sycl::access::mode::read_write; // Mode; // required

    device_iterator() : buffer(cl::sycl::range<1>(1)), idx(0) {}
    device_iterator(sycl::buffer<T, 1/*, Allocator*/> vec, std::size_t index) : buffer(vec), idx(index) {}
    device_iterator(const std::size_t n)
      : buffer(sycl::buffer<T, 1/*, Allocator*/>(sycl::range<1>(n))),
	idx(std::size_t{}) {}
                  #if 0
    reference operator*() const
    {
      auto ptr = (const_cast<device_iterator*>(this)->buffer.template get_access<mode>()).get_pointer();
      return *(ptr + idx);
    }
            #endif
            device_iterator
	    operator+(difference_type forward) const {
	      return device_iterator{buffer, idx + forward};
	    }
    difference_type operator-(const device_iterator &it) const {
      return idx - it.idx;
    }
    bool operator==( device_iterator &it) const {
      return *this - it == 0;
    }
    bool operator!=( device_iterator &it) const {
      return *this - it != 0;
    }
    template <typename OtherIterator>
    typename std::enable_if<
      internal::is_hetero_iterator<OtherIterator>::value,
      difference_type>::type
    operator-(const OtherIterator &it) const {
      return idx - std::distance(dpstd::begin(buffer), it);
    }
    std::size_t get_idx() { return idx; } // required

    sycl::buffer<T, 1/*, Allocator*/> get_buffer() { return buffer; } // required
  };

  template<class InputIter1, class InputIter2>
  void stable_sort_by_key(InputIter1 keys_first, InputIter1 keys_last, InputIter2 values_first)
  {
    auto beg1 = dpstd::zip_iterator<dpct::device_iterator<int>,
				    dpct::device_iterator<int>>(
								keys_first, values_first);

    auto beg2 = dpstd::make_zip_iterator<dpct::device_iterator<int>,
					 dpct::device_iterator<int>>(
								     keys_first, values_first);
    static_assert(std::is_same<decltype(beg1), decltype(beg2)>::value, "");
  }
}

int main() {
  const int numsH = 100;
  dpct::device_iterator<int> mapspkeyT(numsH);
  dpct::device_iterator<int> mapspvalT(numsH);
  cl::sycl::queue q;
  q.submit([&](cl::sycl::handler &h) {
      h.single_task<dpstd::zip_iterator<dpct::device_iterator<int>,
					dpct::device_iterator<int>>>([=] {});
    });

  dpct::stable_sort_by_key(mapspkeyT, mapspkeyT + numsH, mapspvalT);

  std::cout << "done.\n";
  return 0;
}

It is when I looked at the AST that I realized that it was thinking were different:
auto beg1 = dpstd::zip_iterator<dpct::device_iterator<int>,
                                                    dpct::device_iterator<int>>(
                                                                                                keys_first, keys_first);
 
|       |     `-CXXConstructExpr 0x1491f71709b0 <line:47:17, line:49:31> 'dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>>':'dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>>' 'void (const dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>> &)' elidable
                        
|       |           `-CXXBindTemporaryExpr 0x1491f71708d0 <line:47:17, line:49:31> 'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>':'dpstd::zip_iterator< ::dpct::device_iterator<int>, ::dpct::device_iterator<int>>' (CXXTemporary 0x1491f71708d0)
 
    auto beg2 = dpstd::make_zip_iterator<dpct::device_iterator<int>,
                                                             dpct::device_iterator<int>>(
                                                                                                     keys_first, keys_first);
|       |     `-CXXConstructExpr 0x1491f7174e80 <line:51:17, line:53:36> 'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>':'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>' 'void (const dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>> &)' elidable
 
|       |           `-CXXBindTemporaryExpr 0x1491f7174d70 <line:51:17, line:53:36> 'zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>':'dpstd::zip_iterator<dpct::device_iterator<int>, dpct::device_iterator<int>>' (CXXTemporary 0x1491f7174d70) 

beg1 was being using ::dpct::device_iterator<int> but beg2 with dpct::device_iterator<int>

With this information and after talking to Erich, the suggestion was that the getFullyQualifiedType was not the right function to get the type.

What folder should I be using for the test case?

**Comment by zahiraam** [comment]
Is there a way of attaching in here? I don't see it.

File: clang/test/SemaCXX/kernel-function-type.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply clang/test/SemaCXX/kernel-function-type.cpp:3]: Alright, it feels better now, thanks for the explanation.
Can we some kind of this explanation into the test comments?
I think clang/test/SemaSYCL folder will be better for this test.

File: clang/lib/Sema/SemaSYCL.cpp [hunk 2]
```diff
@@ -2200,7 +2197,7 @@ static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
 static std::string getKernelNameTypeString(QualType T, ASTContext &Ctx,
                                            const PrintingPolicy &TypePolicy) {
 
-  QualType FullyQualifiedType = TypeName::getFullyQualifiedType(T, Ctx, true);
+  QualType FullyQualifiedType = T.getCanonicalType();
```
Comment (erichkeane) [inline clang/lib/Sema/SemaSYCL.cpp:2200]: So I cannot help but wonder if this variable is even necessary anymore...

File: clang/test/CodeGenSYCL/kernelname-enum.cpp [hunk 3]
```diff
@@ -203,8 +203,8 @@ int main() {
 // CHECK: template <> struct KernelInfo<::dummy_functor_4<static_cast<enum_in_anonNS>(1)>>
 // CHECK: template <> struct KernelInfo<::dummy_functor_5<static_cast<no_type_set>(0)>>
 // CHECK: template <> struct KernelInfo<::dummy_functor_6<static_cast<unscoped_enum>(0)>>
-// CHECK: template <> struct KernelInfo<::dummy_functor_7<::no_namespace_int>>
-// CHECK: template <> struct KernelInfo<::dummy_functor_7<::internal::namespace_short>>
-// CHECK: template <> struct KernelInfo<::T1<::T2<static_cast<type_argument_template_enum::E>(0)>>>
-// CHECK: template <> struct KernelInfo<::T1<::T3<::type_argument_template_enum::E>>>
-// CHECK: template <> struct KernelInfo<::dummy_functor_8<::EnumTypeOut, Baz>>
+// CHECK: template <> struct KernelInfo<::dummy_functor_7<no_namespace_int>> {
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/kernelname-enum.cpp:206]: Why did you add the curleys here?

File: sycl/test/regression/kernel_name_inside_sycl_namespace.cpp [hunk 4]
```diff
@@ -15,8 +15,8 @@
 #include <CL/sycl.hpp>
 
 struct A {};
-namespace sycl {
 struct B {};
+namespace sycl {
```
Comment (erichkeane) [inline sycl/test/regression/kernel_name_inside_sycl_namespace.cpp:19]: I don't understand this change...

File: clang/test/CodeGenSYCL/int_header1.cpp [hunk 5]
```diff
@@ -2,15 +2,15 @@
 // RUN: FileCheck -input-file=%t.h %s
 
 // CHECK:template <> struct KernelInfo<class KernelName> {
-// CHECK:template <> struct KernelInfo<::nm1::nm2::KernelName0> {
-// CHECK:template <> struct KernelInfo<::nm1::KernelName1> {
-// CHECK:template <> struct KernelInfo<::nm1::KernelName3<::nm1::nm2::KernelName0>> {
-// CHECK:template <> struct KernelInfo<::nm1::KernelName3<::nm1::KernelName1>> {
-// CHECK:template <> struct KernelInfo<::nm1::KernelName4<::nm1::nm2::KernelName0>> {
-// CHECK:template <> struct KernelInfo<::nm1::KernelName4<::nm1::KernelName1>> {
+// CHECK:template <> struct KernelInfo<class nm1::nm2::KernelName0> {
```
Comment (erichkeane) [inline clang/test/CodeGenSYCL/int_header1.cpp:5]: I don't know if we can change this, @elizabethandrews @premanandrao , do you know?  We should possibly instead ensure that the printing policy does FQN instead?  Do either of you have an idea on how to do this?

Comment (premanandrao) [reply clang/test/CodeGenSYCL/int_header1.cpp:5]: If you are asking about the global namespace scope resolution operator, I believe it is needed to ensure that 'nm1' is not looked up in the relative namespace scope.  @elizabethandrews had to do something similar; I think she found FQN didn't help with this and had to use a loop to insert the leading '::'.

Comment (zahiraam) [reply clang/test/CodeGenSYCL/int_header1.cpp:5]: I think Elizabeth was talking about using a flag?

Comment (elizabethandrews) [reply clang/test/CodeGenSYCL/int_header1.cpp:5]: I had to add a flag to printQualifiedName to emit the scope resolution operator at global scope. I believe Clang already has this support for getFullyQualifiedType.

You might need to add something similar to getCanonicalType.

File: sycl/test/regression/kernel_name_inside_sycl_namespace.cpp (hunk 4 above)
Comment (zahiraam) [reply sycl/test/regression/kernel_name_inside_sycl_namespace.cpp:19]: This test case is compiled with flag "-D" "__SYCL_DISABLE_NAMESPACE_INLINE__".
Without this patch, at the second compile triple = "spir64-unknown-linux-sycldevice", a header file is generated with
the declaration of  B under the SYCL_INLINE_NAMESPACE:
__SYCL_INLINE_NAMESPACE(cl) {
	...
template <> struct KernelInfo<::sycl::B> {
	 ...
  }
template <> struct KernelInfo<::A> {
  }
}

At the 3rd compile, this generated header file is included and every thing compiles nice.


With this patch, the header file generated has these declarations for A and B.
template <> struct KernelInfo<struct sycl::B> {
...
}
template <> struct KernelInfo<struct A> {
	 ...	 
}
With this the 3rd compile fails with:
In file included from <built-in>:1:
/tmp/kernel_name_inside_sycl_namespace-8b9eeb.h:44:44: error: no struct named 'B' in namespace 'cl::sycl'
template <> struct KernelInfo<struct sycl::B> {
                                     ~~~~~~^
1 error generated.

To fix this I can either make the change that I am proposing which generates this in the header file:
__SYCL_INLINE_NAMESPACE(cl) {
...			    
template <> struct KernelInfo<struct B> {
	...
  }
};
template <> struct KernelInfo<struct A> {
   ...
}
}
or compile with "-D" "__SYCL_ENABLE_NAMESPACE_INLINE__" .

File: clang/lib/Sema/SemaSYCL.cpp (hunk 2 above)
Comment (zahiraam) [reply clang/lib/Sema/SemaSYCL.cpp]: I have changed that.

File: clang/test/CodeGenSYCL/kernelname-enum.cpp (hunk 3 above)
Comment (zahiraam) [reply clang/test/CodeGenSYCL/kernelname-enum.cpp]: Removed them.

File: sycl/test/regression/kernel_name_inside_sycl_namespace.cpp (hunk 4 above)
Comment (vladimirlaz) [reply sycl/test/regression/kernel_name_inside_sycl_namespace.cpp:19]: In fact the test exposes regression introduced by this patch.
The test code is valid and the compiler should not complain. I would like to block the change until the test passes without change in tests source code.
